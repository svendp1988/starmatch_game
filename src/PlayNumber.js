import React from "react";

export class PlayNumber extends React.Component {
    render() {
        return (
            <button
                className="number"
                onClick={() => this.props.onClick(this.props.number, this.props.status)}
                style={{backgroundColor: colors[this.props.status]}}
            >
                {this.props.number}
            </button>
        )
    }

}

// Color Theme
const colors = {
    available: 'lightgray',
    used: 'lightgreen',
    wrong: 'lightcoral',
    candidate: 'deepskyblue',
};
