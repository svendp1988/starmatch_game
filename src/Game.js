import React from "react";
import "./Game.css";
import {PlayNumber} from "./PlayNumber";
import {StarsDisplay} from "./StarsDisplay";
import {utils} from "./utils";
import PlayAgain from "./PlayAgain";
import {useGameState} from "./useGameState";

const Game = (props) => {
    const numberOfButtons = 9;
    const {
        stars,
        available,
        candidates,
        secondsLeft,
        setGameState
    } = useGameState();

    const determineStatus = (number) => {
        if (!available.includes(number)) {
            return "used";
        }
        if (candidates.includes(number)) {
            return candidatesAreWrong ? "wrong" : "candidate";
        }
        return "available";
    }

    const candidatesAreWrong = utils.sum(candidates) > stars;
    const gameStatus = available.length === 0
        ? "won"
        : secondsLeft === 0
            ? "lost"
            : "active";

    const handleClick = (number, status) => {
        if (gameStatus !== "active" || status === "used") {
            return;
        }
        const newCandidates =
            status === "available"
                ? candidates.concat(number)
                : candidates.filter(candidateNum => candidateNum !== number);
        setGameState(newCandidates);
    }

    return (
        <div className="game">
            <div className="help">
                Pick 1 or more numbers that sum to the number of stars
            </div>
            <div className="body">
                <div className="left">
                    {gameStatus !== "active" ? <PlayAgain onClick={props.resetGame} gameStatus={gameStatus}/> : <StarsDisplay count={stars}/>}
                </div>
                <div className="right">
                    {utils.range(1, numberOfButtons)
                        .map(number =>
                            <PlayNumber
                                key={number}
                                number={number}
                                onClick={handleClick}
                                status={determineStatus(number)}
                            />)}
                </div>
            </div>
            <div className="timer">Time Remaining: {secondsLeft}</div>
        </div>
    )
}

// export class Game extends React.Component {
//     numberOfButtons = 9;
//     state = {
//         stars: utils.random(1, 9),
//         available: utils.range(1, 9),
//         candidates: [],
//         status: ""
//     }
//
//
//     // let stars = generateHtml("div", "star", "");
//
//     determineStatus(number) {
//         if (!this.state.available.includes(number)) {
//             return "used";
//         }
//         if (this.state.candidates.includes(number)) {
//             return this.candidatesAreWrong() ? "wrong" : "candidate";
//         }
//         return "available";
//     }
//
//     candidatesAreWrong() {
//         return utils.sum(this.state.candidates) > this.state.stars;
//     }
//
//     gameIsDone() {
//         return this.state.available.length === 0;
//     }
//
//     handleClick = (number, status) => {
//         console.log("click");
//         if (status === "used") {
//             return;
//         }
//         const newCandidates =
//             status === "available"
//                 ? this.state.candidates.concat(number)
//                 : this.state.candidates.filter(candidateNum => candidateNum !== number);
//         if (utils.sum(newCandidates) !== this.state.stars) {
//             this.setState({
//                 candidates: newCandidates,
//             })
//         } else {
//             const newAvailables = this.state.available.filter(
//                 num => !newCandidates.includes(num)
//             );
//             this.setState({
//                 stars: utils.randomSumIn(newAvailables, 9),
//                 available: newAvailables,
//                 candidates: [],
//             })
//         }
//         // console.log(number + " " + status);
//         // console.log("This: ", this);
//     }
//
//     resetGame = () => {
//         this.setState({
//             stars: utils.random(1, 9),
//             available: utils.range(1, 9),
//             candidates: [],
//         })
//     }
//
//
//
// // function generateHtml(elt, className, txtContent) {
// //     let elts = [];
// //     for (let i = 1; i <= 9; i++) {
// //         elts.push(`<${elt} className=${className}>${txtContent}</${elt}>`);
// //     }
// //     return elts;
// // }
//
// // function generateStars() {
// //     let elts = [];
// //     for (let i = 1; i<= 9; i++) {
// //         elts.push(<div className="star" />);
// //     }
// //     return (elts);
// // }
// //
// // function generateButtons() {
// //     let elts = [];
// //     for (let i = 1; i <= 9; i++) {
// //         elts.push(<button className="number">{i}</button> )
// //     }
// //     return elts;
// // }
//
//     render() {
//         return (
//             <div className="game">
//                 <div className="help">
//                     Pick 1 or more numbers that sum to the number of stars
//                 </div>
//                 <div className="body">
//                     <div className="left">
//                         {/*{stars}*/}
//                         {this.gameIsDone() ? (<PlayAgain onClick={this.resetGame}/>) : (<StarsDisplay count={this.state.stars}/>)}
//                         {/*{utils.range(1, numberOfStars)*/}
//                         {/*    .map(starId => <Star key={starId}/>)}*/}
//
//                         {/*{generateStars()}*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                         {/*<div className="star" />*/}
//                     </div>
//                     <div className="right">
//                         {utils.range(1, this.numberOfButtons)
//                             .map(number =>
//                                 <PlayNumber
//                                     key={number}
//                                     number={number}
//                                     onClick={this.handleClick}
//                                     status={this.determineStatus(number)}
//                                 />)}
//                         {/*{generateButtons()}*/}
//                         {/*<button className="number">1</button>*/}
//                         {/*<button className="number">2</button>*/}
//                         {/*<button className="number">3</button>*/}
//                         {/*<button className="number">4</button>*/}
//                         {/*<button className="number">5</button>*/}
//                         {/*<button className="number">6</button>*/}
//                         {/*<button className="number">7</button>*/}
//                         {/*<button className="number">8</button>*/}
//                         {/*<button className="number">9</button>*/}
//                     </div>
//                 </div>
//                 <div className="timer">Time Remaining: 10</div>
//             </div>
//         )
//     }
// }

export default Game;
