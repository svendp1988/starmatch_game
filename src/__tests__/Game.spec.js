import React from 'react';
import {shallow, mount, spyMethod} from 'enzyme';

import Game from "../Game";
import {utils} from "../utils";
import {render} from "@testing-library/react";
import {PlayNumber} from "../PlayNumber";

describe("Game", () => {
    // snapshot will be different every time because useGameState picks a random number of stars every time.
    it('should match snapshot', function () {
        const container = shallow(<Game/>);
        container.instance().setState({
            stars: 9,
            available: utils.range(1, 9),
            candidates: [],
            secondsLeft: 4,
            gameStatus: "active"
        });
        expect(container.html()).toMatchSnapshot();
    });

    it('should display a helper text', function () {
        const game = shallow(<Game />);
        expect(game.find("div.help").text()).toEqual("Pick 1 or more numbers that sum to the number of stars");
    });

    it('should display a timer with the remaining seconds left', function () {
        const game = shallow(<Game/>);
        expect(game.find("div.timer").text()).toEqual("Time Remaining: 10");
    });

    it('should alter the status on the numbers on clicking it', function () {
        const game = mount(<Game/>);
        const originalStatus = game.find("[number=4]").props().status;
        game.find("[number=4]").simulate("click");
        const statusAfterClick = game.find("[number=4]").props().status;
        expect(originalStatus).not.toEqual(statusAfterClick);
        game.unmount();
    });

    it('should set the status from available to used if number clicked is not available', function () {
        const wrapper = mount(<Game/>);
        expect(wrapper.find("[number=4]").props().status).toEqual("available");
        wrapper.instance().setState({
            available: [1, 2]
        })
        wrapper.find("[number=4]").simulate("click");
        expect(wrapper.find("[number=4]").props().status).toEqual("used");
        wrapper.unmount();
    });

    it('should call the handleClick when a playnumber is clicked', function () {
        const wrapper = mount(<Game />);
        const spy = jest.spyOn(wrapper.instance(), "handleClick");
        wrapper.find("button").first().simulate("click");
        expect(spy).toHaveBeenCalled();
        expect(wrapper.contains(PlayNumber)).toBe(true);
    });

    it('should mock', function () {
        let mockFn = jest.fn();



        const wrapper = mount(<Game />);
        wrapper.instance().handleClick = mockFn;
        console.log(wrapper.find("button").first().props());
        expect(mockFn).toHaveBeenCalledTimes(1);
    });
    /*
    state: {
        stars: 5,
        available: [
          1, 2, 3, 4, 5,
          6, 7, 8, 9
        ],
        candidates: [],
        secondsLeft: 10,
        gameStatus: 'active'
      },
      numberOfButtons: 9,
      setGameState: [Function (anonymous)],
      determineStatus: [Function (anonymous)],
      candidatesAreWrong: false,
      handleClick: [Function (anonymous)],
      setState: [Function (anonymous)],
      myInterval: 2
      */

    it('should give me more insights as to what is going on', function () {
        const wrapper = shallow(<Game/>);
        console.log(wrapper.instance());
    });

    /* in general not sure on how to manipulate the state of this component to control the flow of tests */
})

