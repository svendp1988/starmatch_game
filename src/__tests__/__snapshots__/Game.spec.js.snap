// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should match snapshot 1`] = `"<div class=\\"game\\"><div class=\\"help\\">Pick 1 or more numbers that sum to the number of stars</div><div class=\\"body\\"><div class=\\"left\\"><div class=\\"star\\"></div><div class=\\"star\\"></div><div class=\\"star\\"></div><div class=\\"star\\"></div><div class=\\"star\\"></div><div class=\\"star\\"></div><div class=\\"star\\"></div><div class=\\"star\\"></div><div class=\\"star\\"></div></div><div class=\\"right\\"><button class=\\"number\\" style=\\"background-color:lightgray\\">1</button><button class=\\"number\\" style=\\"background-color:lightgray\\">2</button><button class=\\"number\\" style=\\"background-color:lightgray\\">3</button><button class=\\"number\\" style=\\"background-color:lightgray\\">4</button><button class=\\"number\\" style=\\"background-color:lightgray\\">5</button><button class=\\"number\\" style=\\"background-color:lightgray\\">6</button><button class=\\"number\\" style=\\"background-color:lightgray\\">7</button><button class=\\"number\\" style=\\"background-color:lightgray\\">8</button><button class=\\"number\\" style=\\"background-color:lightgray\\">9</button></div></div><div data-testid=\\"timer\\" class=\\"timer\\">Time Remaining: 4</div></div>"`;
